### PR TITLE
Add  IMA feature to the kernel, add config for it

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

This PR enables the Integrity Measurement Architecture (IMA) in the kernel (in measure mode only). IMA will measure all binaries which are run by root, files which are mmap'd, etc into PCR 10.

IMA does have a performance impact, but enabling the configs should be negligible. Only once a command line parameter is set 
should there be an observable impact.

The feature may be enabled through a new config element `KernelCommandLine` which also allows users to add extra command line parameters in a unified way through the image config file.

###### Change Log  

- Add IMA measurement configs to the x86_64, and aarch64 kernel configs (IMA_APPRAISE currently disabled).
  - Switch some crypto and TPM configs from `m` to `y`
- Add KernelCommandLine config field to control IMA, and allow additional configs to be passed.
  - ImaPolicy key enables IMA features in the kernel
  - ExtraCommandLine appends any additional desired parameters to the end of the kernel command line

###### Does this affect the toolchain?  

NO

###### Test Methodology

- Local builds (x86_64 only), tested on hyper-v with vTPM. IMA measurements are available in `/sys/kernel/security/integrity/ima`

### CODE REVIEW

File: SPECS/kernel/config [hunk 1]
```diff
@@ -2957,7 +2957,7 @@ CONFIG_IPMI_SI=m
 # CONFIG_IPMI_SSIF is not set
```
Comment (christopherco) [inline SPECS/kernel/config:2957]: kernel.spec (and associated `kernel-signed-x64`/`kernel-signed-aarch64`) release number should be updated too.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 2]
```diff
@@ -729,6 +730,29 @@ func InstallGrubCfg(installRoot, rootDevice, bootUUID string, encryptedRoot disk
...
+	}
+
+	// Append any additional command line parameters
+	err = setGrubCfgAdditional(installGrubCfgFile, kernelCommandLine)
+	if err != nil {
+		logger.Log.Warnf("Failed to append extra command line parameterse in grub.cfg: %v", err)
+		return
+	}
+
+	logger.Log.Errorf("Grub at %s", installGrubCfgFile)
```
Comment (christopherco) [inline toolkit/tools/imagegen/installutils/installutils.go:747]: Debugf or Tracef. Applies to the other logger messages below too.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 3]
```diff
@@ -1286,6 +1310,40 @@ func runPostInstallScripts(installChroot *safechroot.Chroot, config configuratio
 	return
 }
 
+func setGrubCfgAdditional(grubPath string, kernelCommandline configuration.KernelCommandLine) (err error) {
+	const (
+		extraPattern = "{{.ExtraCommandLine}}"
+		sedDelimiter = "@"
```
Comment (christopherco) [inline toolkit/tools/imagegen/installutils/installutils.go:1316]: "@" is a valid character in a kernel command line argument (i.e. crashkernel=256M@0). Recommend to use a different delimiter. Maybe ";" or "^"?

**Review by christopherco — CHANGES_REQUESTED** [review]

File: SPECS/kernel/config (hunk 1 above)
Comment (dmcilvaney) [reply SPECS/kernel/config:2957]: Looks like I need to update them all, regardless, I messed up my rebase an lost my spec file change logs as well.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 2 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:747]: Just going to strip them out completely, they are no longer needed.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 3 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:1316]: We use '@' in a lot of places for the command line, I'll udpate everything to use `` ` ``, I think that should be safe?

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 4]
```diff
@@ -729,6 +730,20 @@ func InstallGrubCfg(installRoot, rootDevice, bootUUID string, encryptedRoot disk
...
 
+	// Configure IMA policy
+	err = setGrubCfgIMA(installGrubCfgFile, kernelCommandLine)
+	if err != nil {
+		logger.Log.Warnf("Failed to set ima_policy in grub.cfg: %v", err)
+		return
+	}
+
+	// Append any additional command line parameters
+	err = setGrubCfgAdditional(installGrubCfgFile, kernelCommandLine)
```
Comment (christopherco) [inline toolkit/tools/imagegen/installutils/installutils.go:741]: nit - suggest to update function name for clarity

```suggestion
	err = setGrubCfgAdditionalCmdLine(installGrubCfgFile, kernelCommandLine)
```

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 5]
```diff
@@ -1286,19 +1301,52 @@ func runPostInstallScripts(installChroot *safechroot.Chroot, config configuratio
 	return
 }
 
+func setGrubCfgAdditional(grubPath string, kernelCommandline configuration.KernelCommandLine) (err error) {
```
Comment (christopherco) [inline toolkit/tools/imagegen/installutils/installutils.go:1304]: matching suggestion for nit

```suggestion
func setGrubCfgAdditionalCmdLine(grubPath string, kernelCommandline configuration.KernelCommandLine) (err error) {
```

**Review by christopherco — APPROVED** [review]
Thanks for making the requested changes. Just had a minor nitpick but the rest looks good to me.

**Comment by christopherco** [comment]
Since this also touches some LVM code, adding @niontive as a reviewer

**Review by niontive — APPROVED** [review]

**Pull request merged**